doc: add more context about security in the pipeline and operator

sxd opened the pull request:
The current documentation doesn't cover all the security scans and measures taken in the last year, so we added that to the documentation and also a little section explaining how the operator permissions are managed in the different ways the operator supports and some recommendations about how to deploy it. In the new section, we also explain more about the cluster-wide permissions that are mandatory for the operator to work.

Closes #4348

sxd committed:
doc: add more context about security in the pipeline and operator

The current documentation doesn't cover all the security scan and measures
that has been taken in the last year, so we added that in the documentation
and also, a little section explaining how are the operator permissions
managed in the different way the operator supports and some
recommendations about how to deploy it. In the new section we also
explain more about he cluster wide permissions that are mandatory for the
operator to work.

Closes #4348

Signed-off-by: Jonathan Gonzalez V <jonathan.gonzalez@enterprisedb.com>

mnencia commented on docs/src/security.md:
```diff
@@ -367,3 +388,66 @@ For further detail on how `pg_ident.conf` is managed by the operator, see the
+`nodes` objects, all the other permissions can be namespaced or cluster wide.
+
+The reason for these permissions are:
+* namespace: required to watch other namespaces to keep track of the
+  the objects.
```
```suggestion
* namespace: required to handle the deletion of namespaces containing Clusters.
```

mnencia commented on docs/src/security.md:
```diff
@@ -367,3 +388,66 @@ For further detail on how `pg_ident.conf` is managed by the operator, see the
+The reason for these permissions are:
+* namespace: required to watch other namespaces to keep track of the
+  the objects.
+* nodes: required to get the list of nodes and know the state of them so we avoid
+  scheduling any pod in a node that can't be used.
```
```suggestion
* nodes: required to switch to a different instance if the primary is on a node being cordoned or drained. Without this, the PodDistructionBudget would keep the primary Pod from being evicted, preventing the maintenance operation from being completed.
```

jsilvela commented on docs/src/security.md:
```diff
@@ -26,6 +26,25 @@ a linter that scans the abstract syntactic tree of the source against a set of r
+we run the Snyk code scan every night in a scheduled job and produce weekly reports with any
+new finding on code security and also for licensing issues that could come up.
+
+One of the last things to note is that we have the "Report a vulnerability" option enabled
+on the Security section inside the repository that allows users to safely report any security
```
I don't get this. We do have documentation explaining how to report vulnerabilities.
Do you mean, that section? If so I'll link here.
Been looking at the repo on GH and I don't see any vulnerability reporting tooling.

jsilvela commented on docs/src/security.md:
```diff
@@ -367,3 +388,66 @@ For further detail on how `pg_ident.conf` is managed by the operator, see the
+
+In the default installation for a vanilla Kubernetes cluster, the operator is deployed
+using the manifest that is build using `Kustomize` on the yaml files generated
+by `controller-gen` using the decorations available on [Kubebuilder](https://book.kubebuilder.io/),
```
"decorations" ...???  Do you mean declarations? Customizations? Or is declaration a part of the lingo, like Ansible "playbooks"?

jsilvela committed:
chore: initial round of corrections

Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>

sxd committed:
review

Co-authored-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>
Signed-off-by: Jonathan Gonzalez V. <jonathan.abdiel@gmail.com>
Signed-off-by: Jonathan Gonzalez V <jonathan.gonzalez@enterprisedb.com>

jsilvela committed:
review: further clarifications

Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>

jsilvela commented on docs/src/security.md:
```diff
@@ -367,3 +394,69 @@ For further detail on how `pg_ident.conf` is managed by the operator, see the
+In the default installation for a vanilla Kubernetes cluster, the operator is
+deployed using the manifest built using `Kustomize` on the yaml files generated
+by `controller-gen` using the markers available on [Kubebuilder](https://book.kubebuilder.io/).
+Sadly, these is a bit limited for now, and we can create a manifest
+with all permissions for RoleBinding or ClusterBinding, in our case, the permissions
+are cluster wide in the default deployment, thus, we recommend to deploy the operator
```
I don't understand what you mean here.

jsilvela commented on docs/src/security.md:
```diff
@@ -367,3 +394,69 @@ For further detail on how `pg_ident.conf` is managed by the operator, see the
+to "watch" one, a few or all the namespaces depending on how it is deployed, this open
+the possibility to be more granular on the permissions.
+
+When using OLM the default set of permissions is set to be the two types needed,
+RoleBinding and ClusterBinding, meaning that a set of permissions will be namespaced
+and other permissions will be cluster wide.
```
This is confusingly expressed.
Do you mean that by default, OLM restricts permissions as much as possible? I.e. namespaced where possible?

tijeshjose001z6h commented on docs/src/security.md:
```diff
@@ -367,3 +394,69 @@ For further detail on how `pg_ident.conf` is managed by the operator, see the
+
+The reason for these mandated permissions are:
+
+- namespace: required to handle the deletion of namespaces containing Clusters.
```
The reason for the namespace doesn't seem justified. Could you please elaborate and provide more clarity on this topic, as it is critical from a security perspective?

sxd committed:
chore: removing namespace permissions mention

Signed-off-by: Jonathan Gonzalez V <jonathan.gonzalez@enterprisedb.com>

gbartolini committed:
fix: review

I have also temporarily removed the section about cluster-wide permissions as I think it serves better in a separate patch.

Signed-off-by: Gabriele Bartolini <gabriele.bartolini@enterprisedb.com>

gbartolini commented:
I have temporarily removed the entire section about cluster-wide permissions. I am creating a separate ticket about it, as I think it contradicts - or at least it is not entirely aligned with - the "RBAC" section. I also think that, instead of being thrown at the bottom of the page, it should sit under or next to the RBAC section (which assumes we deploy with the manifest).
